src/cow-react/modules/limitOrders/containers/RateInput/index.tsx

```diff
@@ -0,0 +1,86 @@
+import * as styledEl from './styled'
+import { useCallback, useEffect } from 'react'
+import { RefreshCw } from 'react-feather'
+import { useAtomValue, useUpdateAtom } from 'jotai/utils'
+
+import { HeadingText } from '@cow/modules/limitOrders/pure/RateInput/HeadingText'
+import { limitRateAtom, updateLimitRateAtom } from '@cow/modules/limitOrders/state/limitRateAtom'
+import { limitOrdersAtom } from '@cow/modules/limitOrders/state/limitOrdersAtom'
+import { useCalculateRate } from '@cow/modules/limitOrders/hooks/useCalculateRate'
+import { useUpdateCurrencyAmount } from '@cow/modules/limitOrders/hooks/useUpdateCurrencyAmount'
+import usePrevious from '@src/hooks/usePrevious'
+
+export function RateInput() {
+  const calculateRate = useCalculateRate()
+  const updateCurrencyAmount = useUpdateCurrencyAmount()
+
+  // Rate state
+  const limitRateState = useAtomValue(limitRateAtom)
+  const updateLimitRateState = useUpdateAtom(updateLimitRateAtom)
+  const { isInversed, activeRate, isLoading, marketRate } = limitRateState
+  const prevIsInversed = usePrevious(isInversed)
+
+  // Limit order state
+  const limitOrderState = useAtomValue(limitOrdersAtom)
+  const { inputCurrencyId, outputCurrencyId, inputCurrencyAmount } = limitOrderState
+
+  const primaryCurrency = isInversed ? outputCurrencyId : inputCurrencyId
+  const secondaryCurrency = isInversed ? inputCurrencyId : outputCurrencyId
+
+  // Handle set market price
+  const handleSetMarketPrice = useCallback(() => {
+    updateLimitRateState({ activeRate: marketRate })
+  }, [marketRate, updateLimitRateState])
+
+  // Handle rate input
+  const handleUserInput = useCallback(
+    (typedValue: string) => {
+      updateLimitRateState({ activeRate: typedValue })
+    },
+    [updateLimitRateState]
+  )
+
+  // Handle toggle primary field
+  const handleToggle = () => {
+    const newRate = calculateRate()
+    updateLimitRateState({ isInversed: !isInversed, activeRate: newRate })
+  }
+
+  // Handle rate change but not rate inverse
+  useEffect(() => {
+    if (isInversed === prevIsInversed) {
+      updateCurrencyAmount({ inputCurrencyAmount })
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [activeRate])
```

Comment:
I added a better explanation in the code but: 

We want to observe activeRate change and re-apply the new active rate on the OUTPUT field. To do that we can just call updateCurrencyAmount with the same INPUT value and the new rate will be applied on the OUTPUT field and the new OUTPUT value will be shown.

Now the `isInversed === prevIsInversed` part is so we omit the rate re-calculation if the user click on inverse rate button because in that case we don't want to recalculate the output value, its not needed.

Reply:
As you can see in [useUpdateCurrencyAmount](https://github.com/cowprotocol/cowswap/blob/d848e6fac01dc6aa9cec691ebc932e75191b9277/src/cow-react/modules/limitOrders/hooks/useUpdateCurrencyAmount.ts#L22) if we call this method with inputCurrencyAmount param then the new rate is calculated on the INPUT value and the new OUTPUT value is set, which is what we want to do when the active rate is changed